File: app/api/src/Astra.Api/Llm/Archetypes/java-spring/canonical-php-catalog-service/src/main/java/com/acme/catalog/CouponService.java
// SPDX-Spec: php/discount.php (signed)
// SPDX-Archetype: canonical-php-catalog-service
package com.acme.catalog;

import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Java 21 projection of PHP {@code discount.php}'s {@code applyCoupon}. The PHP
 * source used loose {@code ==} comparisons that mis-handle coupon codes:
 * {@code $couponCode == false} is true for the string {@code "0"}, and (pre-PHP-8)
 * {@code $couponCode == 0} is true for ANY non-numeric string. Both silently void
 * a valid coupon.
 *
 * <p>LTC-1: the migration replaces the juggling with an EXPLICIT typed check —
 * a coupon is "absent" only when the code is null or blank, and codes are
 * matched by value ({@code ===} semantics). The string {@code "0"} is a real,
 * non-empty code and is NOT treated as false.
 */
@TargetMapping(value = "@Service", phpConstruct = "discount.php applyCoupon")
public final class CouponService {

    private static final BigDecimal SAVE10_RATE = new BigDecimal("0.10");

    /**
     * @return the discounted subtotal. A null/blank code applies no discount;
     *         "SAVE10" applies 10%; any other non-blank code applies none.
     */
    @SpecClaim("LTC-1")
    @SpecClaim("INV-1")
    public BigDecimal applyCoupon(BigDecimal subtotal, String couponCode) {
        if (subtotal == null) {
            throw new IllegalArgumentException("subtotal is required");
        }
        // LTC-1: "absent" is null-or-blank ONLY — the string "0" is a real code
        // and must NOT be juggled to false the way PHP's == did.
        if (couponCode == null || couponCode.isBlank()) {
            return subtotal;
        }
        // === semantics: match by value/type, not a loose juggle.
        BigDecimal rate = "SAVE10".equals(couponCode) ? SAVE10_RATE : BigDecimal.ZERO;
        BigDecimal discounted = subtotal.subtract(subtotal.multiply(rate));
        return discounted.setScale(2, RoundingMode.HALF_UP);
    }
}

File: app/api/src/Astra.Api/Llm/Archetypes/dotnet8/canonical-php-catalog-service/src/CouponService.cs
// SPDX-Spec: php/discount.php (signed)
// SPDX-Archetype: canonical-php-catalog-service (dotnet8)
namespace Acme.Catalog;

/// <summary>
/// C# projection of PHP discount.php's applyCoupon. The PHP source used loose
/// <c>==</c> that mis-handles coupon codes ("0" == false is true; pre-PHP-8
/// 0 == "SAVE10" is true). LTC-1: the migration uses an EXPLICIT typed check — a
/// coupon is absent only when null/whitespace, and codes match by value; the
/// string "0" is a real, non-empty code and is NOT treated as false.
/// </summary>
[TargetMapping("service class (DI-registered)", PhpConstruct = "discount.php applyCoupon")]
public sealed class CouponService
{
    private const decimal Save10Rate = 0.10m;

    [SpecClaim("LTC-1")]
    [SpecClaim("INV-1")]
    public decimal ApplyCoupon(decimal subtotal, string? couponCode)
    {
        // LTC-1: "absent" is null-or-whitespace ONLY — "0" is a real code.
        if (string.IsNullOrWhiteSpace(couponCode))
            return subtotal;

        decimal rate = couponCode == "SAVE10" ? Save10Rate : 0m; // === semantics
        decimal discounted = subtotal - (subtotal * rate);
        return Math.Round(discounted, 2, MidpointRounding.AwayFromZero);
    }
}
